src/core/or/connection_edge.c

```diff
@@ -863,18 +863,30 @@ export_hs_client_circuit_id(edge_connection_t *edge_conn,
   if (protocol != HS_CIRCUIT_ID_PROTOCOL_HAPROXY)
     return;
 
+  const origin_circuit_t *circ = NULL;
   char *buf = NULL;
   const char dst_ipv6[] = "::1";
   /* See RFC4193 regarding fc00::/7 */
-  const char src_ipv6_prefix[] = "fc00:dead:beef:4dad:";
+  const char src_ipv6_prefix[] = "fc00:";
   uint16_t dst_port = 0;
   uint16_t src_port = 1; /* default value */
   uint32_t gid = 0; /* default value */
+  uint32_t rp_addr = 0;
+  uint16_t rp_port = 0;
 
   /* Generate a GID and source port for this client */
   if (edge_conn->on_circuit != NULL) {
-    gid = TO_ORIGIN_CIRCUIT(edge_conn->on_circuit)->global_identifier;
+    circ = CONST_TO_ORIGIN_CIRCUIT(edge_conn->on_circuit);
+    gid = circ->global_identifier;
     src_port = gid & 0x0000ffff;
+    if (get_options()->HiddenServiceExportRendPoint) {
+      extend_info_t *extend_info = circ->build_state->chosen_exit;
+      if (BUG(extend_info == NULL)) {
+        goto cleanup;
+      }
+      rp_addr = tor_addr_to_ipv4h(&extend_info->addr);
```

Comment:
This code will fail on IPv6 single onion services, please check that the address is IPv4 before converting.
(As an IPv6 fallback, you can try looking up the IPv4 address in the consensus, but it won't always be there, because the client and service can have different consensuses.)

Reply:
Code removed in `5399de4`.